### PR TITLE
Fixed NSColor dispose side effect.

### PR DESCRIPTION
**Pull request by mkrueger**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1455310
Bug 1455310: [Feedback] Editor theme is not correctly updated when the theme changes (VS for Mac 2022 Preview 4)

### CODE REVIEW

**Comment by mkrueger** [comment]
Seems that NSColor got disposed for the IDE. First exception shown when switching the editor theme:


  System.ObjectDisposedException: Cannot access a disposed object.
  Object name: 'AppKit.NSColor'.
     at ObjCRuntime.ThrowHelper.ThrowObjectDisposedException(Object o) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/ObjCRuntime/ThrowHelper.cs:line 34
     at ObjCRuntime.NativeObjectExtensions.GetNonNullHandle(INativeObject self, String argumentName) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/ObjCRuntime/INativeObject.cs:line 29
     at AppKit.NSBox.set_FillColor(NSColor value) in /Users/builder/azdo/_work/1/s/xamarin-macios/src/build/dotnet/macos/generated-sources/AppKit/NSBox.g.cs:line 580
     at Xamarin.PropertyEditing.Mac.DynamicBox.AppearanceChanged() in /Users/mkrueger/work/vsmac/main/external/Xamarin.PropertyEditing/Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs:line 82
     at Xamarin.PropertyEditing.Mac.DynamicBox.set_HostResourceProvider(IHostResourceProvider value) in /Users/mkrueger/work/vsmac/main/external/Xamarin.PropertyEditing/Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs:line 35
     at Xamarin.PropertyEditing.Mac.PropertyEditorPanel.UpdateResourceProvider() in /Users/mkrueger/work/vsmac/main/external/Xamarin.PropertyEditing/Xamarin.PropertyEditing.Mac/PropertyEditorPanel.cs:line 221
     at Xamarin.PropertyEditing.Mac.PropertyEditorPanel.ViewDidChangeEffectiveAppearance() in /Users/mkrueger/work/vsmac/main/external/Xamarin.PropertyEditing/Xamarin.PropertyEditing.Mac/PropertyEditorPanel.cs:line 213

**Review by Therzok — APPROVED** [review]

**Review by BretJohnson — APPROVED** [review]

**Comment by BretJohnson** [comment]
Approved. @mkrueger - Thanks for this. When you are comfortable with the fix feel free to merge it.

We also just branched PropertyEditing, creating the `d17-0-vsmac` branch which is what should be used for VSMac 17.0. After this is merged to main, we can do the backport and VSMac bump.

**Comment by sevoku** [comment]
@BretJohnson we'd like to take the fix asap, but the build is red here. Please verify and merge if you think it's good to go. Also I don't know how the instertion process goes nowedays, please help if you can.

**Comment by BretJohnson** [comment]
Here's the build error:
```
  /Users/runner/work/1/s/Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs(46,35): error CS0029: Cannot implicitly convert type 'AppKit.NSColor' to 'CoreGraphics.CGColor' [/Users/runner/work/1/s/Xamarin.PropertyEditing.Mac/Xamarin.PropertyEditing.Mac.csproj]
```
@mkrueger - were you able to test this change locally?

File: Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs [hunk 1]
```diff
@@ -56,7 +56,7 @@ public string BorderColorName
 			{
 				this.borderColor = value;
 				if (value == null)
-					BorderColor = NSColor.Clear;
+					this.Layer.BorderColor = NSColor.Clear;
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs:59]:
```suggestion
					this.Layer.BorderColor = NSColor.Clear.CGColor;
```

File: Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ public string FillColorName
 			{
 				this.fillColor = value;
 				if (value == null)
-					FillColor = NSColor.Clear;
+					this.Layer.BackgroundColor = NSColor.Clear;
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs:46]:
```suggestion
					this.Layer.BackgroundColor = NSColor.Clear.CGColor;
```

**Comment by BretJohnson** [comment]
@mkrueger - Can you test this locally, if you haven't already, to confirm it fixes it. I think you probably did test it, but with the build errors I wasn't sure. After confirmation it works, one of us can merge / backport to d17-0-vsmac branch / bump submodule ref in vsmac.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs (hunk 1 above)
Comment (mkrueger) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs]: fixed - changed that after I commited :/

File: Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs (hunk 2 above)
Comment (mkrueger) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/DynamicBox.cs]: fixed

**Comment by mkrueger** [comment]
@BretJohnson y fixes the issue for me locally.

**Comment by sevoku** [comment]
@BretJohnson yeah Mike tested this, but committed the suggestion here on GH afterwards. That should be fixed now and it builds and work locally, however the CI is still red.

**Comment by sevoku** [comment]
And you're right, we can just bump the submodule.

**Review by netonjm — COMMENTED** [review]
@mkruege this is a problem of colors disposed? What's the difference between use NSLayer Background color and NSView Background color?

I'm not sure but I have feelings we need to keep in memory the colors from the HosResourceProviders to don't allow the GC collect the net objects https://github.com/xamarin/vsmac/blob/f55e5639f86a7b6570aea4cfab2942cb88596858/main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport/MonoDevelopHostResourceProvider.cs#L62-L68

**Review by BretJohnson — APPROVED** [review]

**Comment by BretJohnson** [comment]
@gitbot backport d17-0-vsmac

**Comment by Therzok** [comment]
The problem is actaully in Xamarin.Mac https://github.com/xamarin/xamarin-macios/issues/13921

**Pull request merged**